### PR TITLE
Remove the old make-based build system

### LINKED ISSUE #2503: Rewrite build system in Meson
**Checklist in this comment: https://github.com/NixOS/nix/issues/2503#issuecomment-2207393570**

The specific features I am thinking that Autoconf lacks are 
 - Windows support (with and without mingw)
 - Outputs Ninja or something similar

Both of these are non essential, but good for various experiments:
 - Windows support obviously allows experimenting with Nix supporting windows. GCC is *inadequate* because it doesn't offer C++ interopt with MSVC, so Windows's many C++-only APIs are unusable.
 - Outputing a dependency graph allows fun experiments with Recursive Nix (https://github.com/edolstra/nix/commit/1cfdfd728aa7faa78b55b97a2c2f2b5922cd079f) and ninja2nix (https://github.com/awakesecurity/ninja2nix)

This lowers the list down to [Meson](http://mesonbuild.com/) and [CMake](https://cmake.org/), AFIAK. I prefer Meson because:
 - It doesn't use compiler makefile generation, which avoids a regrettable ninja impurity
 - It's less crufty. CMake became better, but old cmake was bad, leaving a lot of legacy cruft. Meson is a clean slate so there's no minefield of deprecated bad ways of doing things.
 - Full disclosure, I've been working on improving Meson's cross compilation support, so I'm partial to it.

CC @taktoa

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Closes #2503

# Context

depends on #11803 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Spooky VM failures?

**Comment by cole-h** [comment]
I don't know if this is already done on `master` --  feel free to tell me it is and I'll remove my comment -- but before this gets merged could we make sure the manual mentions how to actually use this new meson-based build system...?

I've tried to use the new meson way in the past and it was just so much easier to use the make-based build system because it was actually documented in the manual (i.e. how to get a `compile_commands.json` for LSP, how to run specific tests in `tests/`, how to disable the slow documentation generation to be able to iterate quickly, etc).

**Comment by Ericson2314** [comment]
@cole-h I thought the documentation is updated by now, but if it is not do let me know. I am happy to fix that first before merging this.

**Comment by cole-h** [comment]
Is is indeed updated, thanks for making me double check :)

**Comment by edolstra** [comment]
This should also be blocked on https://github.com/NixOS/nix/issues/11741.

**Review by Mic92 — APPROVED** [review]

File: flake.nix
Comment (roberth) [inline flake.nix:221]: or perhaps a tested `nix-cli`, like `nix-everything` but without the libs or manual?

We should probably also add a `nix` attribute with some suggestions of what to pick instead.

**Comment by edolstra** [comment]
Ouch, isn't this premature? I think we decided we want at least one release where the new build system is fully functional while the old one is still available. I really wanted to do a 2.25 release in the next couple of days without having to deal with potential build system fallout.

**Comment by Mic92** [comment]
> Ouch, isn't this premature? I think we decided we want at least one release where the new build system is fully functional while the old one is still available. I really wanted to do a 2.25 release in the next couple of days without having to deal with potential build system fallout.

Ok. I wasn't aware, that this decision was made. We did the merge yesterday when @Ericson2314 @tomberek and @roberth was around.

**Comment by Ericson2314** [comment]
@edolstra My understanding was that we wouldn't *block* the release on this being done, not that we would hold this on waiting for the release.

Are you fine releasing the commit just before this landed? Or are there some other things you want to land first?

**Comment by Mic92** [comment]
Otherwise, I don't think it's a big loss to do a revert, release it and revert the revert. Than have @roberth push his meson nix changes to nixpkgs and we have some time to test nixVersions.nix.

**Comment by goyalyashpal** [comment]
> Otherwise, I don't think it's a big loss to do a revert, release it and revert the revert

https://github.com/NixOS/nix/pull/11850

**Pull request merged**
